consensus/abci.go

```diff
@@ -58,96 +71,174 @@ func NewTendermintABCI(db *database.Database, ethRPCClient *rpc.Client) (*Tender
+	abci.logger.Debug("Initializing chain", "chain_id", req.ChainId)
 
-	return tmtAbciTypes.ResponseInfo{
-		Data:             "ABCIEthereum",
-		LastBlockHeight:  height.Int64(),
-		LastBlockAppHash: root[:],
-	}
+	return tmtAbciTypes.ResponseInitChain{}
 }
 
-// SetOption sets a configuration option
-func (abci *TendermintABCI) SetOption(req tmtAbciTypes.RequestSetOption) tmtAbciTypes.ResponseSetOption {
-	abci.logger.Debug(fmt.Sprintf("Setting option key '%s' value '%s'", req.Key, req.Value))
-	return tmtAbciTypes.ResponseSetOption{Code: tmtAbciTypes.CodeTypeOK, Log: ""}
-}
+// BeginBlock signals the beginning of a new block.
+//
+// The header contains the height, timestamp, and more - it exactly matches the Tendermint block header.
+// The `req.LastCommitInfo` and `req.ByzantineValidators` attributes can be used to determine rewards and punishments
+// for the validators.
+//
+// Response:
+//		- Optional Key-Value tags for filtering and indexing
+func (abci *TendermintABCI) BeginBlock(req tmtAbciTypes.RequestBeginBlock) tmtAbciTypes.ResponseBeginBlock {
+	abci.logger.Debug("Beginning new block", "hash", req.Hash)
+	abci.db.UpdateBlockState(&req.Header)
 
-// InitChain initializes the validator set
-func (abci *TendermintABCI) InitChain(req tmtAbciTypes.RequestInitChain) tmtAbciTypes.ResponseInitChain {
-	abci.logger.Debug(fmt.Sprintf("Initializing chain with ID '%s'", req.ChainId))
-	return tmtAbciTypes.ResponseInitChain{}
+	return tmtAbciTypes.ResponseBeginBlock{}
 }
 
-// CheckTx checks a transaction is valid but does not mutate the state
+// CheckTx validates a mempool transaction, prior to broadcasting or proposing.
+//
+// CheckTx should perform stateful but light-weight checks of the validity of
+// the transaction (like checking signatures and account balances), but need
+// not execute in full (like running a smart contract).
+//
+// The application should maintain a separate state to support CheckTx.
+// This state can be reset to the latest committed state during Persist.
+// Before calling Persist, Tendermint will lock and flush the mempool, ensuring
+// that all existing CheckTx are responded to and no new ones can begin.
+// After Persist, the mempool will rerun CheckTx for all remaining transactions,
+// throwing out any that are no longer valid. Then the mempool will unlock
+// and start sending CheckTx again.
+//
+// Response:
+//		- Response code
+//		- Optional Key-Value tags for filtering and indexing
 func (abci *TendermintABCI) CheckTx(txBytes []byte) tmtAbciTypes.ResponseCheckTx {
 	tx, err := decodeRLP(txBytes)
 	if err != nil {
 		abci.logger.Error("Received invalid transaction", "tx", tx.Hash().String())
 		return tmtAbciTypes.ResponseCheckTx{Code: uint32(abciTypes.ErrEncodingError.Code), Log: err.Error()}
 	}
 
-	return abci.validateTx(tx)
+	abci.logger.Info("Checking TX", "hash", tx.Hash().String(), "nonce", tx.Nonce(), "cost", tx.Cost())
```

Comment:
Why did you move the whole encapsulation of `validatTx` into `checkTx`. Checking a tx and validating a tx does no necessarily be the same. Therefore I would rather keep what we used to have or if you want to optimize you could have created "static" methods for:
- Validate Tx Size
- Validate Sender
- Validate Balance
- Validate Gas
....

That encapsulate would likely clarify this method and also reduce the complexity of reading a 80 lines method

Reply:
because it should never been split in first place. The whole point of CheckTx() is to validate it. They are not different concepts and it's just harder to read having it in a separate method I have to constantly jump to checkTx() that is empty and then to validateTx() to actually see what we are checking. 

Having separate functions for every validation could be nice, although would make the file super long as u would still have to check for an error but create ticket for it please. I want to get this merge finally :D